### ISSUE TITLE
DepDrop with Pagination or Load More

### ISSUE THREAD

**Issue by buttflattery** [kartik-v/yii2-widget-depdrop#68]
Hello Kartik, Good day. i have been trying to figure out working with DepDrop Yii2 plugin and all is good until i came to a situation where i need to limit the data results returned by the response when the parent dropdown onchange is triggered. I want to DepDrop to behave similarly to the select2 `Load more..` / Pagination functionality.

I dont want to ask any usage question here but not sure if it falls into the category or not, as i could not find anything related to the pagination of the results in the DepDrop documentation.

SO, a few things i need to ask

- Firstly is it supported?
- Do we need to set the pagination var for the ajax call using `pluginOptions` under `select2Options` for the DepDrop ? or the `pluginOptions` for the DepDrop where we specify the ajax `url` to fetch the data, as for the `select2` we need to configure the ajax call using the code 

```
'ajax' => [
            'url' => $url ,
            'dataType' => 'json' ,
            'data' => new JsExpression ( 'function(params) { return {q:params.term, page:params.page || 1}; }' )
        ] ,
```

- What should be the response json structure as the normal structure to load data in DepDrop follows the following format

```
{
   "out":[
      {
         "id":1,
         "text":"city1"
      },
      {
         "id":2,
         "text":"city2"
      }
   ],
   "selected":1
}
```
And looking at the Select2 Docs for Load More / Pagination results, we need to return the results as 

```
{
   "results":[
      {
         "id":1,
         "text":"Option 1"
      },
      {
         "id":2,
         "text":"Option 2"
      }
   ],
   "pagination":{
      "more":true
   }
}
```

Currently, i am initializing the DepDrop with the following options where i am passing the `page` parameter via `params` option, 

[
    'id' => 'bill-city',
    'type' => DepDrop::TYPE_SELECT2,
    'pluginOptions' => [
        'placeholder' => 'Select City',
        'depends' => ['state'],
        'url' => Url::to(['/site/cities-list']),
        'params' => ['page'],
    ],
    'options' => [
        'class' => 'form-control',
    ],
],

and formatted the results in the same way that Select2 Pagination section expects, on changing the parent Drop-down it does fetches the results correctly but the DepDrop isnt enabled or populated, it statys disabled and no errors on console, which gives me the feeling that the results formatted in this way wont render the options inside the DepDrop.

How should i go about it if it can be done, **if it falls into the usage questions, ignore it but do add a link to the docs where i can follow.**

Thanks.

**Comment by buttflattery** [comment]
So far i havent been able to find out any solution using the `DepDrop` so i decided to use the normal `Select2` to be used as a dependent drop down.

Anyone looking for the same thing can benefit from this solution.

I have a `state` dropdown and a `city` dropdown and i want to show the cities against the selected state.

State Dropdown

```
$form->field($model,'state')->widget(Select2::class, [
    'data' => ArrayHelper::map(
        $states,
        'code',
        'name'
    ),
    'options' => [
        'placeholder' => 'Select State',
        'class' => 'form-control',
        'id' => 'state',
    ],
    'theme' => Select2::THEME_BOOTSTRAP,
    'pluginOptions' => [
        'allowClear' => true,
    ],
    'pluginEvents' => [
        'select2:select' => 'function(e){enableCities()}',
        'select2:unselect' => 'function(){disableCities()}',
    ],
],)
```
City Dropdown

```
$form->field($model,'city')->widget(Select2::class,[
    'options' => [
        'placeholder' => 'Select City',
        'class' => 'form-control',
        'id' => 'bill-city',
        'disabled' => true,
    ],
    'theme' => Select2::THEME_BOOTSTRAP,
    'pluginOptions' => [
        'allowClear' => true,
        'ajax' => [
            'url' => Url::to(['/site/cities-list']),
            'dataType' => 'json',
            'data' => new JsExpression(
                'function(params){return {q:params.term,page:params.page || 1,stateCode:$("#state").val()};}'
            ),
        ],
    ],

],)
```
Now as you can see i have used the option `"disabled"=>true` for the city drop-down so that it is disabled when the page loads, and in the `state` drop-down you can see that i have a used 2 `pluginEvents` the `select2:select` and `select2:unselect` which are caling the `enableCities()` and `disableCities()` respecitvely which do the same thing as the name describes. So add the following javascript code on top of your view 

```
$js = <<< JS

    function enableCities(){
        $("#bill-city").prop("disabled", false);
    }

    function disableCities(){
        $.fn.select2.defaults.reset();
        $("#bill-city").prop("disabled", true);
    }
JS;
$this->registerJs($js,View::POS_READY);
```
Now you need to implement the server-side code which is the same as the `Select2` suggests i am passing and extra paramter `stateCode` with the params for the ajax call settings in the `city` drop-down so that when you click the drop-down for the first time after selecting the states all th cities relevant to the selcted states are loaded with `load more...` as you scroll down and if you type in. see the below code for reference

```
/**
     * Displays the cities listing for the dropdown select2.
     *
     * @param int    $page      The state code to be searched
     * @param string $q         The city name search term
     * @param string $stateCode The State Code to load cities against
     *
     * @return string
     * @throws NotFoundHttpException
     */
    public function actionCitiesList($page, $q = null, $stateCode = null)
    {
        $request = Yii::$app->request;
        $response = ['results' => ['id' => '', 'text' => '']];

        $isValid = $stateCode !== null;

        if ($isValid) {
            $limit = 15;
            $offset = ($page - 1) * 5;

            Yii::$app->response->format = Response::FORMAT_JSON;

            $query = new Query();
            $query->select('name as id, name as text')
                ->from('{{%city}}')
                ->andFilterWhere(['like', 'name', $q])
                ->andWhere(['=', 'state_code', $stateCode])
                ->limit($limit)
                ->offset($offset)
                ->orderBy('name');

            $command = $query->createCommand();
            $data = $command->queryAll();

            $response['results'] = array_values($data);
            $response['pagination'] = ['more' => !empty($data) ? true : false];
            return $response;
        }
    }
```

**Comment by buttflattery** [comment]
@Xgamefactory yes it should have been `$limit` instead of  `5` or any other static value.

**Comment by buttflattery** [comment]
thanks for confirmation @kartik-v , by the way i get 502 bad gateway on the demo link you provided

**Comment by kartik-v** [comment]
The link works fine.. Can you recheck?

**Comment by buttflattery** [comment]
it wasn't working last night when I checked, i was at home, but in office, the link works fine. The difference may be the ISP as in office we use Transworld which has a foreign registered DNS whereas at home its has the local(Pakistan) registered DNS. I will check again today from home and will update. @kartik-v

**Comment by buttflattery** [comment]
its working fine, never mind.
